algorithmic_efficiency/workloads/wmt/wmt_pytorch/workload.py

```diff
@@ -0,0 +1,357 @@
+"""WMT workload implemented in PyTorch."""
+import contextlib
+from typing import Tuple
+
+from absl import logging
+import jax.dlpack
+import numpy as np
+import tensorflow as tf
+import torch
+import torch.nn.functional as F
+
+from algorithmic_efficiency import spec
+from algorithmic_efficiency.workloads.wmt import bleu
+from algorithmic_efficiency.workloads.wmt import decode
+from algorithmic_efficiency.workloads.wmt.wmt_pytorch.models import Transformer
+from algorithmic_efficiency.workloads.wmt.workload import BaseWmtWorkload
+
+DEVICE = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
+
+
+def jax_to_pytorch(x: spec.Tensor, take_ownership: bool = False):
+  return torch.utils.dlpack.from_dlpack(
+      jax.dlpack.to_dlpack(x, take_ownership=take_ownership))
+
+
+def pytorch_to_jax(x: spec.Tensor):
+  x = x.contiguous()  # https://github.com/google/jax/issues/8082
+  return jax.dlpack.from_dlpack(torch.utils.dlpack.to_dlpack(x))
+
+
+def delete_decoding_cache(model):
+  for module in model.modules():
+    if len(list(module.buffers(recurse=False))) > 0:
+      names = [name for name, _ in module.named_buffers() if name != 'pe']
+      # Deletion has to be done in seperate for-loop to avoid
+      # mutating the OrderedDict during iteration.
+      for name in names:
+        del module._buffers[name]
+
+
+class CrossEntropyLoss(torch.nn.CrossEntropyLoss):
+
+  def forward(self, logits, targets, label_smoothing=0.1):
+    vocab_size = logits.shape[-1]
+    confidence = 1.0 - label_smoothing
+    low_confidence = (1.0 - confidence) / (vocab_size - 1)
+    normalizing_constant = -(
+        confidence * np.log(confidence) +
+        (vocab_size - 1) * low_confidence * np.log(low_confidence + 1e-20))
+    one_hot_targets = F.one_hot(targets, num_classes=vocab_size)
+    soft_targets = torch.where(one_hot_targets == 1, confidence, low_confidence)
+    loss = super().forward(
+        input=logits.transpose(-2, -1), target=soft_targets.transpose(-2, -1))
```

Comment:
I have only created a separate class to inherit from `nn.CrossEntropyLoss` because `DataParallel` is typically applied to modules, although I have just noticed that there is also a [function](https://pytorch.org/docs/stable/generated/torch.nn.functional.torch.nn.parallel.data_parallel.html) for that. It doesn't work with `DistributedDataParallel` though so I will leave this as a separate module for now and revisit it when I move away from `DataParallel`.